ext/crypto/import_key.rs

```diff
@@ -0,0 +1,581 @@
+use std::cell::RefCell;
+use std::rc::Rc;
+
+use deno_core::error::AnyError;
+use deno_core::OpState;
+use deno_core::ZeroCopyBuf;
+use serde::Deserialize;
+use serde::Serialize;
+use spki::der::Decodable;
+use spki::der::Encodable;
+
+use crate::shared::*;
+use crate::OaepPrivateKeyParameters;
+use crate::PssPrivateKeyParameters;
+
+#[derive(Deserialize)]
+#[serde(rename_all = "lowercase")]
+pub enum KeyData {
+  Spki(ZeroCopyBuf),
+  Pkcs8(ZeroCopyBuf),
+  Raw(ZeroCopyBuf),
+  Jwk { k: String },
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase", tag = "algorithm")]
+pub enum ImportKeyOptions {
+  #[serde(rename = "RSASSA-PKCS1-v1_5")]
+  RsassaPkcs1v15 { hash: ShaHash },
+  #[serde(rename = "RSA-PSS")]
+  RsaPss { hash: ShaHash },
+  #[serde(rename = "RSA-OAEP")]
+  RsaOaep { hash: ShaHash },
+  #[serde(rename = "ECDSA", rename_all = "camelCase")]
+  Ecdsa { named_curve: EcNamedCurve },
+  #[serde(rename = "ECDH", rename_all = "camelCase")]
+  Ecdh { named_curve: EcNamedCurve },
+}
+
+#[derive(Serialize)]
+#[serde(untagged)]
+pub enum ImportKeyResult {
+  #[serde(rename_all = "camelCase")]
+  Rsa {
+    raw_data: RawKeyData,
+    modulus_length: usize,
+    public_exponent: ZeroCopyBuf,
+  },
+  #[serde(rename_all = "camelCase")]
+  Ec { raw_data: RawKeyData },
+}
+
+pub async fn op_crypto_import_key(
+  _state: Rc<RefCell<OpState>>,
+  opts: ImportKeyOptions,
+  key_data: KeyData,
+) -> Result<ImportKeyResult, AnyError> {
+  match opts {
+    ImportKeyOptions::RsassaPkcs1v15 { hash } => {
+      import_key_rsassa(key_data, hash)
+    }
+    ImportKeyOptions::RsaPss { hash } => import_key_rsapss(key_data, hash),
+    ImportKeyOptions::RsaOaep { hash } => import_key_rsaoaep(key_data, hash),
+    ImportKeyOptions::Ecdsa { named_curve }
+    | ImportKeyOptions::Ecdh { named_curve } => {
+      import_key_ec(key_data, named_curve)
+    }
+  }
+}
```

Comment:
Why is it an async function is there are no `await` points?

Reply:
Because that is how it already was :shrug:. I'll go make it sync.